### ISSUE TITLE
You actually can't have this code be MIT license with PyQt - it needs to be GPL.

### ISSUE THREAD

**Issue by prakashanurag** [vent4us/vent-dev#5]
You actually can't have this code be MIT license with PyQt - it needs to be GPL.

_Originally posted by @erichulser in https://github.com/vent4us/vent-dev/pull/4_

Convert the License to GPL

**Comment by hongquanli** [comment]
I haven't read much about licenses, but if we want to use MIT instead of GPL, it seems we can do so by switching from PyQt5 to PySide2. In the program, instead directly using PyQt5 I was using qtpy, which allows us to switch from PyQt5 to PySide2 by replacing os.environ["QT_API"] = "pyqt5" with os.environ["QT_API"] = "pyside2".

**Comment by prakashanurag** [comment]
In two lines: GPL will make any further development on this code base to be
submitted back or in turn be open sourced as well in GPL in summary.  MIT
license does not have that limitation.

**Comment by erichulser** [comment]
So, in order to avoid this situation we used PySide vs. PyQt (this was a number of years ago).  That proved to be a really bad choice.

The history of Qt/PyQt is such that it began with Trolltech, was sold to Riverbank (which created PyQt under the GPL license), was sold to Nokia (which created PySide under the LGPL license), was sold to the now Qt foundation.  Because of this constant moving and shifting, the PySide library became abandoned and their bindings [1] were much less stable than PyQt's (it had a shorter life span).

PySide2 at least looks like it is the officially supported Python binding for Qt 5, which in turn is supported by the Qt Company now - which I would hope means that it fixed some of the stability issues.  It looks like it has a lot more activity [2] though it still seems to have a lot of the same problems (262 open tickets) [3], particularly around threading with Python & Qt [4] [5].

I would strongly caution against situations that go frequently back & forth between the GUI / C++ layer (paintEvent, threading) and come up with a different polling model.  From a stability standpoint we abandoned Qt + Python all together because of our frequent crashes & memory leaks, granted we had a high stress environment (dynamically reconfiguring the application for each car coming down the manufacturing line with a 90 second tact time).

[1] https://doc.qt.io/qtforpython/shiboken2/
[2] https://code.qt.io/cgit/pyside/pyside-setup.git/
[3] https://bugreports.qt.io/projects/PYSIDE/issues/PYSIDE-1278?filter=allopenissues
[4] https://bugreports.qt.io/browse/PYSIDE-1108
[5] https://bugreports.qt.io/browse/PYSIDE-973

**Comment by hongquanli** [comment]
Thanks for the detailed note, Eric! And thanks for bringing up the license issue, Anurag! While we probably should stick to Qt5 + Python for the first 10 prototypes for testing and getting clinical feedbacks, it looks like we'll need to make a decision on how to proceed down the road soon. We can discuss this further in the software working document and/or in a software focused zoom meeting.

**Comment by prakashanurag** [comment]
Yep, it's not a problem to move away from pyQt. I am currently more
inclined to get some integration on track with some bare bone models in
place before doing high scale development. We are not at a point of no
return and pretty flexible. Whatever the decision choice is later is
totally fine with me. I don't have much background on this subject of
python libraries and did what I was told. Moving to websockets and
interfacing with some other technology will be fine with me. I will
document whatever I have. It's not that big a code base and should not be
to either, so should be pretty agile and modular. I suppose that should
stay as the intent throughout.

**Comment by hongquanli** [comment]
While we decide on an implementation for the long term, I think at least for the coming 2 weeks we’ll be rolling with your PyQt5 implementation! Lots of data to collect as we may get access to an industry gold standard lung simulator.

**Comment by erichulser** [comment]
The reality is the farther you go down one path the more technical debt you accrue - especially if you think to yourself "this is a temporary path".  The cost of rebuilding is high and will likely not happen.  You would be better served to make a call quickly and move forward.

It may represent duplicate effort but if you think it would be valuable I could port what exists over to a Python backend / React frontend as a comparison if you like.  It would probably take me about a day or two.

**Comment by ethanjli** [comment]
I agree with Eric's point about technical debt - *especially* when we're feeling the time pressure to run "just one more" time-sensitive experiment (and add "just a few more features" in order to carry out that experiment) before dealing with deeper structural issues with the code.

Even after we collect data from the new lung simulator, the need to collect even more new data will only increase, and it will be even more tempting to continue with what we already have. I think the current phase of our project, where we're bringing folks from Brown onboard and critically reviewing design choices made for our prototype, is the best time to decide and execute on the best structure for the GUI software.

**Comment by hongquanli** [comment]
I agree too that based on merits and compliance requirements we should make a decision on the framework to be used sooner than later.
